### PR TITLE
Add `no-unnecessary-qualifier` rule

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [X] Addresses an existing issue: #1987
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [X] Documentation update
- [X] Enable CircleCI for your fork (https://circleci.com/add-projects)

#### What changes did you make?

Added the `no-unnecessary-qualifier` rule, which warns for `A.x` if just `x` would have the same meaning.

### CODE REVIEW

File: src/language/rule/rule.ts [hunk 1]
```diff
@@ -111,6 +111,10 @@ export class Replacement {
         return replacements.reduce((text, r) => r.apply(text), content);
     }
 
+    public static deleteFromTo(start: number, end: number): Replacement {
```
Comment (nchen63) [inline src/language/rule/rule.ts]: use the one in ruleWalker.ts

**Comment by nchen63** [comment]
@andy-hanson  @adidahiya how about naming this `shortest-qualifier`?

**Comment by andy-hanson** [comment]
The rule is about notifying you that a certain node in your AST can be removed if it's unnecessary. It doesn't measure the length of anything.
If you had a rule that warned on `!!x` (where `x` is of type boolean), you wouldn't call it `shortest-boolean-expression`, you'd call it `no-unnecessary-bang-bang`, right?
Open to name changes, but `shortest` implies that it compares multiple possible qualifiers for length.

**Comment by nchen63** [comment]
My concern is that `no-unnecessary-qualifier` sounds like a binary rule, like either have a qualifier or don't have a qualifier. For example, if you're in the `a.b.c` namespace, it would only warn if the qualifier exactly matches `a.b.c`. Maybe it's because I don't view `a` as a qualifier for `b.c` (but should?) I don't feel that strongly about changing the name either way.

**Comment by andy-hanson** [comment]
In my case, I'm just trying to eliminate single qualifiers -- `X.foo` in namespace `X`.
But in #1987 there were multiple qualifiers and a few, but not all, could be removed.

I can see how `no-unnecessary-qualifier` sounds like it does nothing for the latter case. But `shortest-qualifier` sounds like it does nothing for the former case.

Will think about this later.

**Comment by ajafff** [comment]
This should work for enum too. At least the name implies that. Don't know if it does already, but there is no test for it. 

```typescript
enum Foo {
    Bar, 
    Baz = Foo.Bar // could be written as Baz = Bar
}
```

**Comment by nchen63** [comment]
@andy-hanson thanks! name is ok

**Pull request merged**